### PR TITLE
refactor(air): optimize output agg for bitwise

### PR DESCRIPTION
**Pull request by DeimosTheMisthios**
Addresses #256

### CODE REVIEW

File: air/src/aux_table/bitwise_pow2/bitwise/mod.rs [hunk 1]
```diff
@@ -279,6 +308,10 @@ trait EvaluationFrameExt<E: FieldElement> {
     fn bit_decomp(&self) -> &[E];
     /// Gets the entire range of decomposed input values for `a` and `b` in the next row.
     fn bit_decomp_next(&self) -> &[E];
+    /// Gets the value of the aggregated output in the previous row.
+    fn output_prev(&self) -> E;
+    /// Gets the value of the aggregated output in the next row of the previous row.
```
Comment (grjte) [inline air/src/aux_table/bitwise_pow2/bitwise/mod.rs:313]: I think "in the next row of the previous row" is a bit unclear. Can you rephrase for clarity?

File: docs/src/design/aux_table/bitwise.md [hunk 2]
```diff
@@ -88,20 +91,22 @@ The above constraints enforce that when $k_1 = 1$ , $a' = 16 \cdot a + \sum_{i=0
...
+$$
 
+Next, we need to ensure that for each row except the last, the aggregated output value must equal the previous aggregated output value in the next row.
 $$
-k_0 \cdot \left(z - \sum_{i=0}^3(2^i \cdot a_i \cdot b_i)\right) = 0
+k_1 \cdot \left(z - z'_p\right) = 0
 $$
 
-Lastly, we need to ensure that for all other rows, the value in the $z$ column is computed by multiplying the value from the previous row of the column by 16 and then adding it to the bitwise operation applied to the next set of bits of $a$ and $b$. This can be enforced with the following constraint:
+Lastly, we need to ensure that for all rows, the value in the $z$ column is computed by multiplying the previous value (from the current row) by 16 and then adding it to the bitwise operation applied to the row's set of bits of $a$ and $b$. This can be enforced with the following constraint:
```
Comment (grjte) [inline docs/src/design/aux_table/bitwise.md:104]: nit: It might be a bit clearer to say "... by multiplying the previous output value (from the $z_p$ column in the current row)..."

**Review by grjte — CHANGES_REQUESTED** [review]
Looks good, thank you! I left a few nits inline as well as a performance-related comment.

**Comment by DeimosTheMisthios** [comment]
Thanks @grjte for the helpful review. Please let me know if there's anything else we need here.

File: air/src/aux_table/bitwise_pow2/bitwise/mod.rs [hunk 3]
```diff
@@ -279,6 +270,11 @@ trait EvaluationFrameExt<E: FieldElement> {
     fn bit_decomp(&self) -> &[E];
     /// Gets the entire range of decomposed input values for `a` and `b` in the next row.
     fn bit_decomp_next(&self) -> &[E];
+    /// Gets the value of the aggregated output in the previous row.
+    fn output_prev(&self) -> E;
+    /// Gets the value of the aggregated output of the current row, or
+    /// the previous row with respect to the next row.
+    fn output_prev_next(&self) -> E;
```
Comment (grjte) [inline air/src/aux_table/bitwise_pow2/bitwise/mod.rs:277]: Since we're no longer using `bit_decomp_next` (right above this), it can be removed. The same is true for `output_next` (below).

File: air/src/aux_table/bitwise_pow2/bitwise/mod.rs [hunk 4]
```diff
@@ -151,12 +149,13 @@ fn enforce_input_decomposition<E: FieldElement>(
...
 /// for each operation:
-/// - In the first row, the output value must equal the aggregated result of the operation applied
-///   to the row.
-/// - For every row except the last, the output value in the next row must equal the output in the
-///   current row plus the aggregated result of the specified operation applied to the next row.
+/// - In the first row, the aggregated output value of the previous row must be zero.
+/// - For every row except the last, the aggregated output value must equal the previous aggregated
+///   output value in the next row.
+/// - For every row, the output value must equal 16 times the previous output plus the aggregated
+///   result of the specified operation.
```
Comment (grjte) [inline air/src/aux_table/bitwise_pow2/bitwise/mod.rs:158]: nit: Sorry I missed this last time, but I think the output aggregation constraint descriptions used in the `pow2` module are a bit clearer (they had more eyes on them). Since this is doing the same thing maybe we should just reuse those descriptions here:

```
/// - In the first row, `output_prev` should be set to 0.
/// - For all the rows except the last one, the next value of `output_prev` should be the same as
///   the current value of `output`.
/// - For all rows except the last one, the current output value (`output`) should equal the output value
///   copied from the previous row (`output_prev`) plus the aggregation of current row's decomposed power.
```

**Review by grjte — CHANGES_REQUESTED** [review]
Thank you! I left a couple more small things inline, then we'll be good to go here.

**Review by grjte — APPROVED** [review]
Looks good to me! Thanks @DeimosTheMisthios!
